File: src/custom_msgs/src/py/cm_sensor.py
#!/usr/bin/env python
import rospy
from rospy.exceptions import ROSInterruptException
import random

# Import custom messages
from custom_msgs.msg import IoTSensor

# Define node actions
def node():
	# Initialize the node
	rospy.init_node("cm_sensor", anonymous=False)

	# Create a publisher
	pub = rospy.Publisher("cm_data", IoTSensor, queue_size=10)

	# Keep publishing new data
	rate = rospy.Rate(1)
	i = 0
	while not rospy.is_shutdown():
		# Create a random data point
		data = IoTSensor()
		data.id = i
		data.name = "IoT Sensor data"
		data.temperature = random.randint(0, 256)
		data.humidity = random.randint(0, 100)

		# Publish it
		rospy.loginfo(data)
		pub.publish(data)
		rate.sleep()
		i += 1

# Start the node when the file is executed
if __name__ == "__main__":
	try:
		node()
	except ROSInterruptException:
		pass

File: src/custom_msgs/src/cpp/cm_sensor.cpp
#include <ros/ros.h>
#include <random>
#include <ctime>

// Include custom messages
#include "custom_msgs/IoTSensor.h"

void log_data(const custom_msgs::IoTSensor& data_point) {
  ROS_INFO("id: %d\nname: %s\ntemperature: %lf\nhumidity: %lf\n",
         data_point.id,
         data_point.name.c_str(),
         data_point.temperature,
         data_point.humidity);
}

int random_int(int max) {
  return std::rand() % max;
}

// Define node actions
int main(int argc, char* argv[]) {
  std::srand(std::time(NULL));

  // Initialize node
  ros::init(argc, argv, "cm_sensor");

  // Create a publisher
  ros::NodeHandle node;
  ros::Publisher pub = node.advertise<custom_msgs::IoTSensor>("cm_data", 10);

  // Keep publishing new data
  ros::Rate rate(1);
  int i = 0;
  while (ros::ok()) {
    // Create random data point
    custom_msgs::IoTSensor data;
    data.id = i;
    data.name = "IoT Sensor data";
    data.temperature = random_int(255);
    data.humidity = random_int(100);

    // Publish it
    log_data(data);
    pub.publish(data);
    ros::spinOnce();
    rate.sleep();
    ++i;
  }

  return 0;
}
